File: ui/config_test.go
package ui_test

import (
	"math"
	"strings"
	"testing"
	"time"

	"github.com/stretchr/testify/require"

	"github.com/miroslav-matejovsky/adsb-testbench/simulatorapi"
	"github.com/miroslav-matejovsky/adsb-testbench/ui"
)

// managerConfig returns a complete explicit fixture configuration.
func managerConfig() ui.ManagerConfig {
	return ui.ManagerConfig{
		APIBaseURL: "/api/simulator/", AssetBaseURL: "/assets/",
		ManagerSettings: ui.ManagerSettings{
			PollIntervalMilliseconds: 1000, RequestTimeoutMilliseconds: 5000,
			MaxResponseBytes: 1 << 20, ResumeSpeedHundredths: 100,
		},
	}
}

// aircraftConfig returns a complete explicit fixture configuration.
func aircraftConfig() ui.AircraftDisplayConfig {
	return ui.AircraftDisplayConfig{
		APIBaseURL: "/api/display/", AssetBaseURL: "/assets/",
		AircraftDisplaySettings: ui.AircraftDisplaySettings{
			PollIntervalMilliseconds: 1000, RequestTimeoutMilliseconds: 5000,
			MaxResponseBytes: 1 << 20, StationIDs: []string{"alpha"},
			FreshFor: 10 * time.Second, LostAfter: 60 * time.Second,
			HistoryPageSize: 50, MaxHistoryRecords: 500,
			InitialLatitudeDegrees: 50, InitialLongitudeDegrees: 14, InitialZoom: 7,
			Tiles: nil,
		},
	}
}

func tiles() *ui.Tiles {
	return &ui.Tiles{
		URLTemplate:     "https://tiles.example.test/{z}/{x}/{y}.png",
		AttributionText: "Example tiles", AttributionURL: "https://tiles.example.test/about",
		MinZoom: 0, MaxZoom: 18,
	}
}

func TestConfigsAcceptCompleteSettings(t *testing.T) {
	t.Parallel()

	require.NoError(t, managerConfig().Validate())
	require.NoError(t, aircraftConfig().Validate())

	withTiles := aircraftConfig()
	withTiles.Tiles = tiles()
	require.NoError(t, withTiles.Validate())

	empty := aircraftConfig()
	empty.StationIDs = []string{}
	require.NoError(t, empty.Validate(), "an empty selection is explicit")

	absolute := managerConfig()
	absolute.APIBaseURL = "http://127.0.0.1:8080/bench/a/api/simulator"
	require.NoError(t, absolute.Validate())
}

func TestManagerConfigRejectsInvalidSettings(t *testing.T) {
	t.Parallel()

	cases := map[string]func(*ui.ManagerConfig){
		"relative api":       func(c *ui.ManagerConfig) { c.APIBaseURL = "api/" },
		"protocol relative":  func(c *ui.ManagerConfig) { c.APIBaseURL = "//evil.test/api/" },
		"script api":         func(c *ui.ManagerConfig) { c.APIBaseURL = "javascript:alert(1)" },
		"backslash asset":    func(c *ui.ManagerConfig) { c.AssetBaseURL = `/assets\x/` },
		"encoded dot asset":  func(c *ui.ManagerConfig) { c.AssetBaseURL = "/a/%2e%2e/assets/" },
		"zero poll":          func(c *ui.ManagerConfig) { c.PollIntervalMilliseconds = 0 },
		"negative timeout":   func(c *ui.ManagerConfig) { c.RequestTimeoutMilliseconds = -1 },
		"small response":     func(c *ui.ManagerConfig) { c.MaxResponseBytes = simulatorapi.MinResponseBytes - 1 },
		"zero resume speed":  func(c *ui.ManagerConfig) { c.ResumeSpeedHundredths = 0 },
		"missing api":        func(c *ui.ManagerConfig) { c.APIBaseURL = "" },
		"query in api":       func(c *ui.ManagerConfig) { c.APIBaseURL = "/api/?x=1" },
		"userinfo api":       func(c *ui.ManagerConfig) { c.APIBaseURL = "http://u:p@host/api/" },
		"control byte asset": func(c *ui.ManagerConfig) { c.AssetBaseURL = "/assets\n/" },
	}
	for name, mutate := range cases {
		t.Run(name, func(t *testing.T) {
			t.Parallel()

			config := managerConfig()
			mutate(&config)
			require.ErrorIs(t, config.Validate(), ui.ErrInvalidConfig)
		})
	}
}

func TestAircraftConfigRejectsInvalidSettings(t *testing.T) {
	t.Parallel()

	cases := map[string]func(*ui.AircraftDisplayConfig){
		"nil stations":       func(c *ui.AircraftDisplayConfig) { c.StationIDs = nil },
		"duplicate station":  func(c *ui.AircraftDisplayConfig) { c.StationIDs = []string{"a", "a"} },
		"invalid station":    func(c *ui.AircraftDisplayConfig) { c.StationIDs = []string{"a b"} },
		"long station":       func(c *ui.AircraftDisplayConfig) { c.StationIDs = []string{strings.Repeat("a", 65)} },
		"too many stations":  func(c *ui.AircraftDisplayConfig) { c.StationIDs = strings.Split("a,b,c,d,e,f,g,h,i", ",") },
		"zero fresh":         func(c *ui.AircraftDisplayConfig) { c.FreshFor = 0 },
		"lost not after":     func(c *ui.AircraftDisplayConfig) { c.LostAfter = c.FreshFor },
		"zero page":          func(c *ui.AircraftDisplayConfig) { c.HistoryPageSize = 0 },
		"large page":         func(c *ui.AircraftDisplayConfig) { c.HistoryPageSize = 1001 },
		"records below page": func(c *ui.AircraftDisplayConfig) { c.MaxHistoryRecords = 49 },
		"nan latitude":       func(c *ui.AircraftDisplayConfig) { c.InitialLatitudeDegrees = math.NaN() },
		"latitude range":     func(c *ui.AircraftDisplayConfig) { c.InitialLatitudeDegrees = 91 },
		"infinite longitude": func(c *ui.AircraftDisplayConfig) { c.InitialLongitudeDegrees = math.Inf(1) },
		"zoom range":         func(c *ui.AircraftDisplayConfig) { c.InitialZoom = 25 },
		"ftp api":            func(c *ui.AircraftDisplayConfig) { c.APIBaseURL = "ftp://host/api/" },
		"tile without z": func(c *ui.AircraftDisplayConfig) {
			c.Tiles = tiles()
			c.Tiles.URLTemplate = "https://tiles.test/{x}/{y}.png"
		},
		"tile repeated x": func(c *ui.AircraftDisplayConfig) {
			c.Tiles = tiles()
			c.Tiles.URLTemplate = "https://tiles.test/{z}/{x}/{x}/{y}.png"
		},
		"tile other placeholder": func(c *ui.AircraftDisplayConfig) {
			c.Tiles = tiles()
			c.Tiles.URLTemplate = "https://{s}.tiles.test/{z}/{x}/{y}.png"
		},
		"tile script": func(c *ui.AircraftDisplayConfig) {
			c.Tiles = tiles()
			c.Tiles.URLTemplate = "javascript:alert('{z}{x}{y}')"
		},
		"tile userinfo": func(c *ui.AircraftDisplayConfig) {
			c.Tiles = tiles()
			c.Tiles.URLTemplate = "https://u@tiles.test/{z}/{x}/{y}.png"
		},
		"tile fragment": func(c *ui.AircraftDisplayConfig) {
			c.Tiles = tiles()
			c.Tiles.URLTemplate = "https://tiles.test/{z}/{x}/{y}.png#a"
		},
		"tile protocol relative": func(c *ui.AircraftDisplayConfig) {
			c.Tiles = tiles()
			c.Tiles.URLTemplate = "//tiles.test/{z}/{x}/{y}.png"
		},
		"empty attribution": func(c *ui.AircraftDisplayConfig) {
			c.Tiles = tiles()
			c.Tiles.AttributionText = " "
		},
		"script attribution link": func(c *ui.AircraftDisplayConfig) {
			c.Tiles = tiles()
			c.Tiles.AttributionURL = "javascript:alert(1)"
		},
		"zoom outside tiles": func(c *ui.AircraftDisplayConfig) {
			c.Tiles = tiles()
			c.Tiles.MinZoom = 8
		},
		"inverted tile zoom": func(c *ui.AircraftDisplayConfig) {
			c.Tiles = tiles()
			c.Tiles.MinZoom = 10
			c.Tiles.MaxZoom = 5
		},
	}
	for name, mutate := range cases {
		t.Run(name, func(t *testing.T) {
			t.Parallel()

			config := aircraftConfig()
			mutate(&config)
			require.ErrorIs(t, config.Validate(), ui.ErrInvalidConfig)
		})
	}
}

func parse(t *testing.T, text string) *simulatorapi.Value {
	t.Helper()

	value, err := simulatorapi.ParseJSONBytes([]byte(text))
	require.NoError(t, err)
	return value
}

const managerSettingsJSON = `{"pollIntervalMilliseconds":1000,"requestTimeoutMilliseconds":5000,` +
	`"maxResponseBytes":1048576,"resumeSpeedHundredths":100}`

const aircraftSettingsJSON = `{"pollIntervalMilliseconds":1000,"requestTimeoutMilliseconds":5000,` +
	`"maxResponseBytes":1048576,"stationIds":["alpha"],"freshForNanoseconds":"10000000000",` +
	`"lostAfterNanoseconds":"60000000000","historyPageSize":50,"maxHistoryRecords":500,` +
	`"initialLatitudeDegrees":50,"initialLongitudeDegrees":14,"initialZoom":7,"tiles":null}`

func TestParseSettingsBindsEveryKey(t *testing.T) {
	t.Parallel()

	manager, err := ui.ParseManagerSettings(parse(t, managerSettingsJSON))
	require.NoError(t, err)
	require.Equal(t, managerConfig().ManagerSettings, manager)

	aircraft, err := ui.ParseAircraftDisplaySettings(parse(t, aircraftSettingsJSON))
	require.NoError(t, err)
	require.Equal(t, aircraftConfig().AircraftDisplaySettings, aircraft)

	withTiles := strings.Replace(aircraftSettingsJSON, `"tiles":null`,
		`"tiles":{"urlTemplate":"https://tiles.example.test/{z}/{x}/{y}.png","attributionText":"Example tiles",`+
			`"attributionUrl":"https://tiles.example.test/about","minZoom":0,"maxZoom":18}`, 1)
	aircraft, err = ui.ParseAircraftDisplaySettings(parse(t, withTiles))
	require.NoError(t, err)
	require.Equal(t, tiles(), aircraft.Tiles)
}

func TestParseSettingsRejectsIncompleteInput(t *testing.T) {
	t.Parallel()

	managerCases := map[string]string{
		"missing key":  strings.Replace(managerSettingsJSON, `,"resumeSpeedHundredths":100`, ``, 1),
		"null value":   strings.Replace(managerSettingsJSON, `"resumeSpeedHundredths":100`, `"resumeSpeedHundredths":null`, 1),
		"unknown key":  strings.Replace(managerSettingsJSON, `{`, `{"extra":1,`, 1),
		"string value": strings.Replace(managerSettingsJSON, `1000`, `"1000"`, 1),
		"fraction":     strings.Replace(managerSettingsJSON, `1000`, `1000.5`, 1),
		"zero speed":   strings.Replace(managerSettingsJSON, `"resumeSpeedHundredths":100`, `"resumeSpeedHundredths":0`, 1),
	}
	for name, text := range managerCases {
		_, err := ui.ParseManagerSettings(parse(t, text))
		require.ErrorIs(t, err, ui.ErrInvalidConfig, name)
	}

	aircraftCases := map[string]string{
		"missing tiles":      strings.Replace(aircraftSettingsJSON, `,"tiles":null`, ``, 1),
		"null stations":      strings.Replace(aircraftSettingsJSON, `["alpha"]`, `null`, 1),
		"numeric duration":   strings.Replace(aircraftSettingsJSON, `"10000000000"`, `10000000000`, 1),
		"padded duration":    strings.Replace(aircraftSettingsJSON, `"10000000000"`, `"010000000000"`, 1),
		"unknown tile key":   strings.Replace(aircraftSettingsJSON, `"tiles":null`, `"tiles":{"x":1}`, 1),
		"fractional zoom":    strings.Replace(aircraftSettingsJSON, `"initialZoom":7`, `"initialZoom":7.5`, 1),
		"unknown key":        strings.Replace(aircraftSettingsJSON, `{`, `{"apiBaseUrl":"/api/",`, 1),
		"thresholds swapped": strings.Replace(aircraftSettingsJSON, `"60000000000"`, `"1"`, 1),
	}
	for name, text := range aircraftCases {
		_, err := ui.ParseAircraftDisplaySettings(parse(t, text))
		require.ErrorIs(t, err, ui.ErrInvalidConfig, name)
	}

	_, err := simulatorapi.ParseJSONBytes([]byte(`{"a":1,"a":2}`))
	require.Error(t, err, "duplicate keys are rejected by the shared strict parser")
}

File: ui/browser/unit/config.test.mjs
// Unit tests for browser configuration validation. The URL vectors mirror
// internal/urlpath so Go and browser code accept and reject the same bases.
import assert from "node:assert/strict";
import { test } from "node:test";

import {
  ConfigError, resolveBase, validateAircraftConfig, validateManagerConfig, validateStationIds,
} from "../../internal/assets/static/config.js";

const origin = "http://127.0.0.1:8080";

function manager(overrides = {}) {
  return {
    apiBaseUrl: "/bench/a/api/simulator/", assetBaseUrl: "/bench/a/assets/",
    pollIntervalMilliseconds: 1000, requestTimeoutMilliseconds: 5000,
    maxResponseBytes: 1048576, resumeSpeedHundredths: 100, ...overrides,
  };
}

function aircraft(overrides = {}) {
  return {
    apiBaseUrl: "/api/display/", assetBaseUrl: "/assets/",
    pollIntervalMilliseconds: 1000, requestTimeoutMilliseconds: 5000, maxResponseBytes: 1048576,
    stationIds: ["alpha"], freshForNanoseconds: "10000000000", lostAfterNanoseconds: "60000000000",
    historyPageSize: 3, maxHistoryRecords: 6, initialLatitudeDegrees: 50, initialLongitudeDegrees: 14,
    initialZoom: 7, tiles: null, ...overrides,
  };
}

const tiles = {
  urlTemplate: "https://tiles.example.test/{z}/{x}/{y}.png", attributionText: "Example",
  attributionUrl: "https://tiles.example.test/about", minZoom: 0, maxZoom: 18,
};

test("resolveBase keeps nested prefixes and adds one trailing slash", () => {
  assert.equal(resolveBase("k", "/", origin), `${origin}/`);
  assert.equal(resolveBase("k", "/bench/a/api/simulator", origin), `${origin}/bench/a/api/simulator/`);
  assert.equal(resolveBase("k", `${origin}/bench/a/assets`, origin), `${origin}/bench/a/assets/`);
  assert.equal(resolveBase("k", "/bench%20a/", origin), `${origin}/bench%20a/`);
});

test("resolveBase rejects ambiguous and cross-origin bases before any request", () => {
  for (const raw of [
    "", "bench/a/", "//evil.test/api/", "/bench//a/", "/bench/./a/", "/bench/../a/", "/bench/%2e%2E/a/",
    "/bench%2Fa/", "/bench%5Ca/", "/bench\\a/", "/\\evil.test/", "/bench/?a=1", "/bench/?", "/bench/#",
    "/bench\n/", "/bench%0A/", "/bench%zz/", "javascript:alert(1)", "data:text/html,x",
    "http://evil.test/api/", "https://127.0.0.1:8080/api/", "http://127.0.0.1:8081/api/",
    "http://user@127.0.0.1:8080/api/", "http://127.0.0.1:8080/a/../api/",
  ]) {
    assert.throws(() => resolveBase("apiBaseUrl", raw, origin), ConfigError, JSON.stringify(raw));
  }
});

test("validateManagerConfig resolves bases and requires every key", () => {
  const config = validateManagerConfig(manager(), origin);
  assert.equal(config.apiBaseUrl, `${origin}/bench/a/api/simulator/`);
  assert.equal(config.resumeSpeedHundredths, 100);
  assert.ok(Object.isFrozen(config));

  for (const key of Object.keys(manager())) {
    const missing = manager();
    delete missing[key];
    assert.throws(() => validateManagerConfig(missing, origin), ConfigError, `missing ${key}`);
    assert.throws(() => validateManagerConfig(manager({ [key]: null }), origin), ConfigError, `null ${key}`);
  }
  assert.throws(() => validateManagerConfig(manager({ extra: 1 }), origin), ConfigError);
  assert.throws(() => validateManagerConfig(manager({ pollIntervalMilliseconds: 0 }), origin), ConfigError);
  assert.throws(() => validateManagerConfig(manager({ requestTimeoutMilliseconds: 1.5 }), origin), ConfigError);
  assert.throws(() => validateManagerConfig(manager({ maxResponseBytes: 2047 }), origin), ConfigError);
  assert.throws(() => validateManagerConfig(manager({ resumeSpeedHundredths: 0 }), origin), ConfigError);
  assert.throws(() => validateManagerConfig(null, origin), ConfigError);
});

test("validateAircraftConfig parses exact thresholds and nullable tiles", () => {
  const config = validateAircraftConfig(aircraft(), origin);
  assert.equal(config.freshForNanoseconds, 10000000000n);
  assert.equal(config.lostAfterNanoseconds, 60000000000n);
  assert.equal(config.tiles, null);
  assert.deepEqual([...config.stationIds], ["alpha"]);

  const withTiles = validateAircraftConfig(aircraft({ tiles }), origin);
  assert.equal(withTiles.tiles.urlTemplate, tiles.urlTemplate);
  const local = validateAircraftConfig(aircraft({ tiles: { ...tiles, urlTemplate: "/tiles/{z}/{x}/{y}.png" } }), origin);
  assert.equal(local.tiles.urlTemplate, "/tiles/{z}/{x}/{y}.png");
  assert.deepEqual([...validateAircraftConfig(aircraft({ stationIds: [] }), origin).stationIds], []);
});

test("validateAircraftConfig rejects out-of-domain settings", () => {
  const invalid = [
    { stationIds: null }, { stationIds: "alpha" }, { stationIds: ["a", "a"] }, { stationIds: ["a b"] },
    { freshForNanoseconds: 10 }, { freshForNanoseconds: "0" }, { lostAfterNanoseconds: "10000000000" },
    { freshForNanoseconds: "010" }, { historyPageSize: 0 }, { historyPageSize: 1001 },
    { maxHistoryRecords: 2 }, { initialLatitudeDegrees: 91 }, { initialLatitudeDegrees: Number.NaN },
    { initialLongitudeDegrees: -181 }, { initialZoom: 7.5 }, { initialZoom: 25 },
    { tiles: { ...tiles, urlTemplate: "https://t.test/{x}/{y}.png" } },
    { tiles: { ...tiles, urlTemplate: "https://{s}.t.test/{z}/{x}/{y}.png" } },
    { tiles: { ...tiles, urlTemplate: "javascript:alert('{z}{x}{y}')" } },
    { tiles: { ...tiles, urlTemplate: "//t.test/{z}/{x}/{y}.png" } },
    { tiles: { ...tiles, urlTemplate: "https://u@t.test/{z}/{x}/{y}.png" } },
    { tiles: { ...tiles, urlTemplate: "https://t.test/{z}/{x}/{y}.png#a" } },
    { tiles: { ...tiles, attributionText: "" } },
    { tiles: { ...tiles, attributionUrl: "javascript:alert(1)" } },
    { tiles: { ...tiles, minZoom: 8 } },
    { tiles: { ...tiles, extra: true } },
    { tiles: { urlTemplate: tiles.urlTemplate } },
  ];
  for (const overrides of invalid) {
    assert.throws(() => validateAircraftConfig(aircraft(overrides), origin), ConfigError, JSON.stringify(overrides));
  }
});

test("validateStationIds enforces the selection bound", () => {
  assert.throws(() => validateStationIds(["a", "b", "c", "d", "e", "f", "g", "h", "i"]), ConfigError);
  assert.deepEqual([...validateStationIds(["a-1", "B_2"])], ["a-1", "B_2"]);
});
